asciidoctorj-core/src/main/java/org/asciidoctor/jruby/internal/JRubyAsciidoctor.java

```diff
@@ -407,6 +400,17 @@ public <T> T convertFile(File file, Map<String, Object> options, Class<T> expect
         }
     }
 
+    private <T> T adaptReturn(IRubyObject object, Class<T> expectedResult) {
+        if (NodeConverter.NodeType.DOCUMENT_CLASS.isInstance(object)) {
+            if (Document.class.isAssignableFrom(expectedResult)) {
+                return (T) new DocumentImpl(object);
```

Comment:
Maybe it would be better to use `NodeConverter.createNode()` if `expectedResult` is Document.class?
Then we would keep NodeConverter as the only class that creates AST nodes.

Reply:
Absolutely!